File: node/src/tools/pnpm.mjs
import { pnpmRcPath } from "../util/paths.mjs";
import { readSafe, writeAtomic } from "../util/io.mjs";
import { readKey, setKey, removeKey } from "../util/lines.mjs";
import { readExtras, applyExtras, removeExtras } from "../util/extras.mjs";
import { detectVersion, gte } from "../util/version.mjs";

export const name = "pnpm";
export const key = "minimum-release-age";
export const docs = "https://pnpm.io/settings#minimumreleaseage";
export const minBin = [10, 6, 0];
// `defaultSinceMajor`: pnpm major version where the value became the default.
// Once detected, an absent line in the rc file is still effectively in force,
// so `read()` reports it as ok with `defaultEnforced: true`.
export const extras = [
  { key: "trust-policy", expected: "no-downgrade", line: "trust-policy=no-downgrade" },
  { key: "block-exotic-subdeps", expected: "true", line: "block-exotic-subdeps=true", defaultSinceMajor: 11 },
  { key: "strict-dep-builds", expected: "true", line: "strict-dep-builds=true" },
  { key: "verify-deps-before-run", expected: "error", line: "verify-deps-before-run=error" },
  { key: "minimum-release-age-strict", expected: "true", line: "minimum-release-age-strict=true" },
];

export function path(ctx) { return pnpmRcPath(ctx.env, ctx.home, ctx.platform); }

// Cache `pnpm --version` for the lifetime of the process. preflight() and
// read() both want it; without memoization a single `pmsec check` spawns
// pnpm twice. Cache key is the override env value so tests with different
// `PMSEC_PNPM_VERSION` settings invalidate naturally.
let _versionCache = null;
function pnpmVersion(ctx) {
  const override = ctx.env?.PMSEC_PNPM_VERSION ?? null;
  if (_versionCache && _versionCache.override === override) return _versionCache.value;
  const v = detectVersion("pnpm", ["--version"], { env: ctx.env, overrideKey: "PMSEC_PNPM_VERSION" });
  _versionCache = { override, value: v };
  return v;
}

export function preflight(ctx) {
  const v = pnpmVersion(ctx);
  if (v === null) return { ok: true, message: null };
  if (gte(v, minBin)) return { ok: true, version: v.raw, message: null };
  return { ok: true, warn: true, version: v.raw, message: `pnpm ${v.raw} < ${minBin.join(".")}: minimum-release-age is silently ignored. Upgrade pnpm to enforce the cooldown.` };
}

export async function read(ctx) {
  const p = path(ctx);
  const raw = await readSafe(p);
  const value = readKey(raw, key);
  const minutes = value === null ? null : Number(value);
  const v = pnpmVersion(ctx);
  const defaults = new Map(extras.filter(x => x.defaultSinceMajor).map(x => [x.key, x.defaultSinceMajor]));
  const evaluated = readExtras(raw, extras).map(e => {
    const def = defaults.get(e.key);
    if (e.configured === null && def && v && v.major >= def) {
      return { ...e, ok: true, defaultEnforced: true };
    }
    return e;
  });
  return {
    path: p, configured: value,
    days: minutes === null || Number.isNaN(minutes) ? null : Math.floor(minutes / (60 * 24)),
    extras: evaluated
  };
}

export async function write(days, ctx) {
  const p = path(ctx);
  let text = setKey(await readSafe(p), key, `${key}=${days * 24 * 60}`);
  text = applyExtras(text, extras);
  await writeAtomic(p, text);
  return { path: p };
}

export async function unset(ctx) {
  const p = path(ctx);
  const before = await readSafe(p);
  const cooldown = removeKey(before, key);
  const ex = removeExtras(cooldown.text, extras);
  const removed = cooldown.removed || ex.removed;
  if (removed) await writeAtomic(p, ex.text);
  return { path: p, removed };
}

File: python/src/pmsec/tools/pnpm.py
from __future__ import annotations

from pathlib import Path

from pmsec.util.context import Context
from pmsec.util.extras import apply_extras, read_extras, remove_extras
from pmsec.util.io import write_atomic
from pmsec.util.lines import read_key, remove_key, set_key
from pmsec.util.paths import pnpm_rc_path
from pmsec.util.version import detect_version, gte

NAME = "pnpm"
KEY = "minimum-release-age"
DOCS = "https://pnpm.io/settings#minimumreleaseage"
MIN_BIN = (10, 6, 0)
# `default_since_major`: pnpm major version where the value became the default.
# When detected, an absent line in the rc file is still effectively in force,
# so `read()` reports it as ok with `defaultEnforced: True`.
EXTRAS = [
    {"key": "trust-policy", "expected": "no-downgrade", "line": "trust-policy=no-downgrade"},
    {"key": "block-exotic-subdeps", "expected": "true", "line": "block-exotic-subdeps=true", "default_since_major": 11},
    {"key": "strict-dep-builds", "expected": "true", "line": "strict-dep-builds=true"},
    {"key": "verify-deps-before-run", "expected": "error", "line": "verify-deps-before-run=error"},
    {"key": "minimum-release-age-strict", "expected": "true", "line": "minimum-release-age-strict=true"},
]


def path(ctx: Context) -> Path:
    return pnpm_rc_path(ctx.env, ctx.home, ctx.platform)


# Cache `pnpm --version` for the lifetime of the process. preflight() and
# read() both want it; without memoization a single `pmsec check` spawns
# pnpm twice. Cache key is the override env value so tests with different
# `PMSEC_PNPM_VERSION` settings invalidate naturally.
_version_cache: tuple[str | None, tuple[int, int, int, str] | None] | None = None


def _pnpm_version(ctx: Context):
    global _version_cache
    override = ctx.env.get("PMSEC_PNPM_VERSION")
    if _version_cache is not None and _version_cache[0] == override:
        return _version_cache[1]
    v = detect_version("pnpm", env=ctx.env, override_key="PMSEC_PNPM_VERSION")
    _version_cache = (override, v)
    return v


def preflight(ctx: Context) -> dict:
    v = _pnpm_version(ctx)
    if v is None:
        return {"ok": True, "message": None}
    if gte(v, MIN_BIN):
        return {"ok": True, "version": v[3], "message": None}
    msg = (
        f"pnpm {v[3]} < {'.'.join(str(n) for n in MIN_BIN)}: "
        "minimum-release-age is silently ignored. Upgrade pnpm to enforce the cooldown."
    )
    return {"ok": True, "warn": True, "version": v[3], "message": msg}


def read(ctx: Context) -> dict:
    p = path(ctx)
    raw = p.read_text("utf-8") if p.exists() else ""
    value = read_key(raw, KEY)
    minutes = None
    if value is not None:
        try:
            minutes = int(value)
        except ValueError:
            minutes = None
    days = None if minutes is None else minutes // (60 * 24)
    v = _pnpm_version(ctx)
    defaults = {e["key"]: e["default_since_major"] for e in EXTRAS if e.get("default_since_major")}
    extras_rows = []
    for row in read_extras(raw, EXTRAS):
        d = defaults.get(row["key"])
        if row["configured"] is None and d and v is not None and v[0] >= d:
            row = {**row, "ok": True, "defaultEnforced": True}
        extras_rows.append(row)
    return {"path": str(p), "configured": value, "days": days, "extras": extras_rows}


def write(days: int, ctx: Context) -> dict:
    p = path(ctx)
    raw = p.read_text("utf-8") if p.exists() else ""
    text = apply_extras(set_key(raw, KEY, f"{KEY}={days * 24 * 60}"), EXTRAS)
    write_atomic(p, text)
    return {"path": str(p)}


def unset(ctx: Context) -> dict:
    p = path(ctx)
    if not p.exists():
        return {"path": str(p), "removed": False}
    before = p.read_text("utf-8")
    after, removed_cooldown = remove_key(before, KEY)
    after, removed_extras = remove_extras(after, EXTRAS)
    removed = removed_cooldown or removed_extras
    if removed:
        write_atomic(p, after)
    return {"path": str(p), "removed": removed}
